server/game/gameSystems/DelayedEffectSystem.ts

```diff
@@ -0,0 +1,93 @@
+import AbilityHelper from '../AbilityHelper';
+import { AbilityContext } from '../core/ability/AbilityContext';
+import { IAbilityLimit } from '../core/ability/AbilityLimit';
+import { TriggeredAbilityContext } from '../core/ability/TriggeredAbilityContext';
+import { Duration, EventName } from '../core/Constants';
+import { GameEvent } from '../core/event/GameEvent';
+import { GameSystem, IGameSystemProperties } from '../core/gameSystem/GameSystem';
+import { WhenType } from '../Interfaces';
+import * as Contract from '../core/utils/Contract';
+
+export enum DelayedEffectType {
+    Card,
+    Player
+}
+
+export interface IDelayedEffectSystemProperties extends IGameSystemProperties {
+    title: string;
+    when: WhenType;
+    duration?: Duration;
+    limit?: IAbilityLimit;
+    immediateEffect: GameSystem<TriggeredAbilityContext>;
+    effectType: DelayedEffectType;
+}
+
+export class DelayedEffectSystem<TContext extends AbilityContext = AbilityContext> extends GameSystem<TContext, IDelayedEffectSystemProperties> {
+    public override readonly name: string = 'applyDelayedEffect';
+    public override readonly eventName: EventName = EventName.OnEffectApplied;
+    public override readonly effectDescription: string = 'apply a delayed effect';
+
+    protected override defaultProperties: IDelayedEffectSystemProperties = {
+        title: null,
+        when: null,
+        duration: Duration.Persistent,
+        limit: AbilityHelper.limit.perGame(1),
+        immediateEffect: null,
+        effectType: null
+    };
+
+    public eventHandler(event: any, additionalProperties: any): void {
+        const properties = this.generatePropertiesFromContext(event.context, additionalProperties);
+        // TODO Remove this if we don't need it
+        // if (!properties.ability) {
+        //     properties.ability = event.context.ability;
+        // }
+
+        const { title, when, duration, limit, immediateEffect, ...otherProperties } = properties;
+
+        const renamedProperties = { ...otherProperties, ongoingEffect:
+            AbilityHelper.ongoingEffects.delayedEffect({
+                title,
+                when,
+                immediateEffect,
+                limit
+            }) };
+
+        const delayedEffectSource = properties.effectType === DelayedEffectType.Card ? event.context.target : event.context.source;
```

Comment:
Was thinking on this some more, I think right now it will fail if used outside of a target resolver. Like if we just have it as an `immediateEffect` instead of `targetResolver` there will be no `event.context.target`.

I think what we actually need is to override `addPropertiesToEvent()`, determine the source card there, and attach it to the event as `event.sourceCard` or something. Then we would reference it here.

We may also want to just put this logic in a separate method so it could be called from other places like `hasLegalTarget`, see comment there